pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,259 @@
+package clustermemberremovalcontroller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	configv1informers "github.com/openshift/client-go/config/informers/externalversions/config/v1"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	machinelistersv1beta1 "github.com/openshift/client-go/machine/listers/machine/v1beta1"
+	"github.com/openshift/cluster-etcd-operator/pkg/dnshelpers"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/operatorclient"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+// errNotFound a sentinel errors indicating a given item hasn't been found
+var errNotFound = errors.New("not found")
+
+type ClusterMemberRemovalController struct {
+	etcdClient                        etcdcli.EtcdClient
+	masterNodeLister                  corev1listers.NodeLister
+	masterMachineLister               machinelistersv1beta1.MachineLister
+	networkLister                     configv1listers.NetworkLister
+	configMapListerForTargetNamespace corev1listers.ConfigMapNamespaceLister
+	// isMachineAPIFunctionalFn determines if the precondition for this controller is met,
+	// this controller can be run only on a cluster that exposes a functional Machine API
+	isMachineAPIFunctionalFn func() (bool, error)
+	masterMachineSelector    labels.Selector
+	masterNodeSelector       labels.Selector
+}
+
+// NewClusterMemberRemovalController removes an etcd member if the machine and a node for the etcd member is gone and the machine-api is active
+//
+// Note:
+//  since this controller needs to reconcile only master nodes and machine objects
+//  make sure nodeInformer and machineInformer contain only filtered data
+//  otherwise it might be expensive to react to every node update in larger installations
+func NewClusterMemberRemovalController(
+	etcdClient etcdcli.EtcdClient,
+	isMachineAPIFunctional func() (bool, error),
+	masterMachineSelector labels.Selector, masterNodeSelector labels.Selector,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	masterNodeInformer cache.SharedInformer,
+	masterMachineInformer cache.SharedInformer,
+	masterNodeLister corev1listers.NodeLister,
```

Comment:
Why is the lister also provided here?  If I have the informer, I have the ability to build the lister with just a `.Lister` which sensures a matched set, right?

Reply:
not quite right, the informer is `cache.SharedInformer` so you have to `corev1listers.NewNodeLister(masterNodeInformer.GetIndexer())`